File: src/LiveAlert.Core/YouTubeLiveDetector.cs
using System.Net.Http.Headers;
using System.Text.RegularExpressions;

namespace LiveAlert.Core;

public interface ILiveDetector
{
    Task<LiveCheckResult> CheckLiveAsync(AlertConfig alert, CancellationToken cancellationToken);
}

public sealed class YouTubeLiveDetector : ILiveDetector
{
    private readonly HttpClient _httpClient;
    private static readonly Regex VideoIdRegex = new("\"videoId\":\"([A-Za-z0-9_-]{11})\"", RegexOptions.Compiled);
    private static readonly Regex IsLiveNowRegex = new("\"isLiveNow\":(true|false)", RegexOptions.Compiled);

    public YouTubeLiveDetector(HttpClient httpClient)
    {
        _httpClient = httpClient;
        if (!_httpClient.DefaultRequestHeaders.UserAgent.Any())
        {
            _httpClient.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue("Mozilla", "5.0"));
        }
    }

    public async Task<LiveCheckResult> CheckLiveAsync(AlertConfig alert, CancellationToken cancellationToken)
    {
        try
        {
            var url = alert.Url.Trim();
            if (url.Contains("watch?v=", StringComparison.OrdinalIgnoreCase))
            {
                var videoId = ExtractVideoIdFromUrl(url);
                if (string.IsNullOrEmpty(videoId))
                {
                    return LiveCheckResult.NotLive();
                }

                return await CheckWatchPageAsync(videoId, cancellationToken).ConfigureAwait(false);
            }

            var liveUrl = BuildLiveUrl(url);
            var (liveOk, liveHtml, liveError) = await FetchStringAsync(liveUrl, cancellationToken).ConfigureAwait(false);
            if (!liveOk)
            {
                return liveError ?? LiveCheckResult.Error("HTTP error");
            }
            var match = VideoIdRegex.Match(liveHtml ?? string.Empty);
            if (!match.Success)
            {
                return LiveCheckResult.NotLive();
            }

            var candidateId = match.Groups[1].Value;
            return await CheckWatchPageAsync(candidateId, cancellationToken).ConfigureAwait(false);
        }
        catch (OperationCanceledException)
        {
            throw;
        }
        catch (Exception ex)
        {
            return LiveCheckResult.Error(ex.Message);
        }
    }

    private async Task<LiveCheckResult> CheckWatchPageAsync(string videoId, CancellationToken cancellationToken)
    {
        var watchUrl = $"https://www.youtube.com/watch?v={videoId}";
        var (watchOk, html, watchError) = await FetchStringAsync(watchUrl, cancellationToken).ConfigureAwait(false);
        if (!watchOk)
        {
            return watchError ?? LiveCheckResult.Error("HTTP error");
        }
        var liveMatch = IsLiveNowRegex.Match(html ?? string.Empty);
        if (liveMatch.Success && liveMatch.Groups[1].Value.Equals("true", StringComparison.OrdinalIgnoreCase))
        {
            return LiveCheckResult.Live(videoId);
        }

        return LiveCheckResult.NotLive();
    }

    private static string BuildLiveUrl(string url)
    {
        if (url.Contains("/live", StringComparison.OrdinalIgnoreCase))
        {
            return url;
        }

        url = url.TrimEnd('/');
        return url + "/live";
    }

    private static string? ExtractVideoIdFromUrl(string url)
    {
        var idx = url.IndexOf("watch?v=", StringComparison.OrdinalIgnoreCase);
        if (idx < 0) return null;
        var part = url[(idx + "watch?v=".Length)..];
        var amp = part.IndexOf('&');
        if (amp >= 0)
        {
            part = part[..amp];
        }
        return part.Length == 11 ? part : null;
    }

    private async Task<(bool ok, string? content, LiveCheckResult? error)> FetchStringAsync(string url, CancellationToken cancellationToken)
    {
        using var response = await _httpClient.GetAsync(url, cancellationToken).ConfigureAwait(false);
        if (!response.IsSuccessStatusCode)
        {
            return (false, null, LiveCheckResult.Error($"HTTP {(int)response.StatusCode} {response.ReasonPhrase}"));
        }

        var html = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
        return (true, html, null);
    }
}

public readonly record struct LiveCheckResult(bool IsLive, string? VideoId, bool IsError, string? ErrorMessage)
{
    public static LiveCheckResult Live(string videoId) => new(true, videoId, false, null);
    public static LiveCheckResult NotLive() => new(false, null, false, null);
    public static LiveCheckResult Error(string message) => new(false, null, true, message);
}

File: src/LiveAlert.Electron/src/core/youtube-live-detector.ts
import { AlertConfig } from './config-models';

export interface LiveCheckResult {
  isLive: boolean;
  videoId?: string;
  isError: boolean;
  errorMessage?: string;
}

const IS_LIVE_NOW_REGEX = /"isLiveNow":(true|false)/;
const INITIAL_DATA_REGEX = /var ytInitialData\s*=\s*(\{.*?\});/s;
const USER_AGENT = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36';

export class YouTubeLiveDetector {
  private abortController: AbortController | null = null;

  async checkLive(alert: AlertConfig, signal?: AbortSignal): Promise<LiveCheckResult> {
    try {
      const url = alert.url.trim();
      if (url.includes('watch?v=')) {
        const videoId = extractVideoIdFromUrl(url);
        if (!videoId) return { isLive: false, isError: false };
        return await this.checkWatchPage(videoId, signal);
      }

      const baseUrl = normalizeChannelUrl(url);
      const streamsUrl = `${baseUrl}/streams`;
      const streamsResult = await this.fetchString(streamsUrl, signal);
      if (streamsResult.error) {
        return { isLive: false, isError: true, errorMessage: streamsResult.error };
      }

      let candidateId = extractLiveVideoId(streamsResult.content);
      if (!candidateId) {
        const homeResult = await this.fetchString(baseUrl, signal);
        if (homeResult.error) {
          return { isLive: false, isError: true, errorMessage: homeResult.error };
        }
        candidateId = extractLiveVideoId(homeResult.content);
      }

      if (!candidateId) {
        return { isLive: false, isError: false };
      }

      return await this.checkWatchPage(candidateId, signal);
    } catch (e: any) {
      if (e.name === 'AbortError') throw e;
      return { isLive: false, isError: true, errorMessage: e.message };
    }
  }

  cancel(): void {
    this.abortController?.abort();
  }

  private async checkWatchPage(videoId: string, signal?: AbortSignal): Promise<LiveCheckResult> {
    const watchUrl = `https://www.youtube.com/watch?v=${videoId}`;
    const result = await this.fetchString(watchUrl, signal);
    if (result.error) {
      return { isLive: false, isError: true, errorMessage: result.error };
    }
    const match = IS_LIVE_NOW_REGEX.exec(result.content ?? '');
    if (match && match[1] === 'true') {
      return { isLive: true, videoId, isError: false };
    }
    return { isLive: false, isError: false };
  }

  private async fetchString(
    url: string,
    signal?: AbortSignal
  ): Promise<{ content?: string; error?: string }> {
    try {
      const response = await fetch(url, {
        headers: { 'User-Agent': USER_AGENT },
        signal,
      });
      if (!response.ok) {
        return { error: `HTTP ${response.status} ${response.statusText}` };
      }
      const content = await response.text();
      return { content };
    } catch (e: any) {
      if (e.name === 'AbortError') throw e;
      return { error: e.message };
    }
  }
}

function normalizeChannelUrl(url: string): string {
  url = url.replace(/\/+$/, '');
  const lowered = url.toLowerCase();
  for (const suffix of ['/streams', '/live', '/videos', '/featured']) {
    if (lowered.endsWith(suffix)) {
      return url.slice(0, -suffix.length);
    }
  }
  return url;
}

function extractVideoIdFromUrl(url: string): string | null {
  const idx = url.toLowerCase().indexOf('watch?v=');
  if (idx < 0) return null;
  let part = url.slice(idx + 'watch?v='.length);
  const amp = part.indexOf('&');
  if (amp >= 0) part = part.slice(0, amp);
  return part.length === 11 ? part : null;
}

function extractLiveVideoId(html?: string): string | null {
  if (!html) return null;
  const match = INITIAL_DATA_REGEX.exec(html);
  if (!match) return null;
  try {
    const data = JSON.parse(match[1]);
    return findLiveVideoId(data);
  } catch {
    return null;
  }
}

function findLiveVideoId(element: any): string | null {
  if (element === null || element === undefined) return null;
  if (Array.isArray(element)) {
    for (const item of element) {
      const found = findLiveVideoId(item);
      if (found) return found;
    }
    return null;
  }
  if (typeof element === 'object') {
    if (element.videoId && typeof element.videoId === 'string' && Array.isArray(element.thumbnailOverlays)) {
      if (hasLiveOverlay(element.thumbnailOverlays)) {
        return element.videoId;
      }
    }
    for (const key of Object.keys(element)) {
      const found = findLiveVideoId(element[key]);
      if (found) return found;
    }
  }
  return null;
}

function hasLiveOverlay(overlays: any[]): boolean {
  for (const overlay of overlays) {
    const renderer = overlay?.thumbnailOverlayTimeStatusRenderer;
    if (!renderer) continue;
    if (typeof renderer.style === 'string' && renderer.style.toUpperCase() === 'LIVE') {
      return true;
    }
    if (renderer.text && textContainsLiveLabel(renderer.text)) {
      return true;
    }
  }
  return false;
}

function textContainsLiveLabel(textElement: any): boolean {
  if (typeof textElement !== 'object') return false;
  if (typeof textElement.simpleText === 'string' && textElement.simpleText.includes('ライブ')) {
    return true;
  }
  if (Array.isArray(textElement.runs)) {
    for (const run of textElement.runs) {
      if (typeof run?.text === 'string' && run.text.includes('ライブ')) {
        return true;
      }
    }
  }
  return false;
}
